BlueprintUICommonControls/Sources/EditingMenu.swift

```diff
@@ -0,0 +1,309 @@
+//
+//  EditingMenu.swift
+//  BlueprintUICommonControls
+//
+//  Created by Kyle Van Essen on 1/5/22.
+//
+
+import BlueprintUI
+
+
+///
+/// Allows showing the system's `UIMenuController` editing menu upon long press of the wrapped element.
+///
+/// ```
+/// myElement.editingMenu {
+///     EditingMenuItem(.copy) {
+///         print("Copied!")
+///     }
+///
+///     EditingMenuItem(.select) {
+///         print("Selected!")
+///     }
+/// }
+/// ```
+public struct EditingMenu: Element {
+
+    /// The wrapped element to display.
+    public var wrapped: Element
+
+    /// The editing items to show in the editing menu.
+    public var items: [EditingMenuItem]
+
+    /// Creates a new editing menu, wrapping the provided element, and displaying the provided items.
+    public init(
+        wrapping: Element,
+        @Builder<EditingMenuItem> with items: () -> [EditingMenuItem]
+    ) {
+        let items = items()
+
+        precondition(items.isEmpty == false, "Must provide at least one `EditingMenuItem` to the `EditingMenu`.")
+
+        wrapped = wrapping
+        self.items = items
+    }
+
+    // MARK: Element
+
+    public var content: ElementContent {
+        ElementContent(child: wrapped)
+    }
+
+    public func backingViewDescription(with context: ViewDescriptionContext) -> ViewDescription? {
+        View.describe { config in
+            config[\.items] = items
+        }
+    }
+}
+
+
+/// A single item in an editing menu.
+public struct EditingMenuItem {
+
+    /// The type of menu item.
+    public var kind: Kind
+
+    /// A callback, invoked when the user selects the menu item.
+    public var onSelect: () -> Void
+
+    /// Creates a new menu item of the given kind.
+    public init(
+        _ kind: Kind,
+        onSelect: @escaping () -> Void
+    ) {
+        self.kind = kind
+        self.onSelect = onSelect
+    }
+}
+
+
+extension EditingMenuItem {
+
+    /// All of the different menu item kinds supported by an editing menu.
+    ///
+    /// ### Note
+    /// Matches methods from `UIResponderStandardEditActions`.
+    public enum Kind {
```

Comment:
Should this be a struct with static methods so you can add custom ones?

Reply:
Soooo the way that the `UIMenuController` API works is really miserable: The actions you can perform are based on the menu traversing the responder chain, and seeing which things in the responder chain handle the provided selector via `canPerformAction` (like `cut:`, `paste:`, etc); and then calls that selector on the responder. Because of this, custom actions would need to pass in a custom object with a custom selector that can get forward by the backing view here – which was a giant pain in the ass, so I decided to just not.

Reply:
Figured out a reasonable-ish way to support this, going to add.